system/src/commonMain/kotlin/com/splendo/kaluga/system/network/state/NetworkStateRepo.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network.state
+
+import co.touchlab.stately.concurrency.AtomicReference
+import com.splendo.kaluga.base.runBlocking
+import com.splendo.kaluga.state.ColdStateRepo
+import com.splendo.kaluga.system.network.BaseNetworkManager
+import com.splendo.kaluga.system.network.Network
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+class NetworkStateRepo(
+    private val networkManagerBuilder: BaseNetworkManager.Builder,
+) : ColdStateRepo<NetworkState>() {
+
+    interface Builder {
+        fun create(): NetworkStateRepo
+    }
+
+    private var _lastKnownNetwork = AtomicReference<Network>(Network.Unknown.WithoutLastNetwork(Network.Unknown.Reason.NOT_CLEAR))
+    internal var lastKnownNetwork: Network
+        get() = _lastKnownNetwork.get()
+        set(value) { _lastKnownNetwork.set(value) }
+
+    internal var networkManager: BaseNetworkManager? = null
+        set(value) {
+            field?.let {
+                if(value == null) {
+                    it.dispose()
+                }
+            }
+            field = value
+        }
+
+    override suspend fun deinitialize(state: NetworkState) {
+        lastKnownNetwork = state.networkType
+        networkManager = null
+    }
+
+    override suspend fun initialValue(): NetworkState {
+        networkManager = networkManagerBuilder.create(::onNetworkStateChange)
+
+        return when(val network = lastKnownNetwork) {
+            is Network.Unknown.WithoutLastNetwork -> NetworkState.Unknown(
+                Network.Unknown.WithoutLastNetwork(network.reason)
+            )
+            is Network.Unknown.WithLastNetwork -> NetworkState.Unknown(
+                Network.Unknown.WithLastNetwork(network.lastKnownNetwork, network.reason)
+            )
+            is Network.Known.Cellular -> NetworkState.Available(Network.Known.Cellular())
+            is Network.Known.Wifi -> NetworkState.Available(
+                Network.Known.Wifi(network.isExpensive)
```

Comment:
Why do you keep assuming because the _last_ known network state (which for all we know was months ago) was something the _current_ network state is something?

Every single branch of this `when` should return some form of `Network.Unknown`

Reply:
I'm pretty sure I commit the changes for this... I guess when I was playing with git reset I deleted that commit and forgot about it. I will push the changes

Reply:
b43a483